create sign in route

Ahmed-aed opened the pull request:
Fixes #19 and fixes #23
- complete sign-up route.
- update signup route.
-  create authentication middleware in the server and use them to pass clientId to the protected client route.
- update the schema and fake data to include default avatar and role.
- update password validation to include uppercase and special characters.
-  add morgan middleware to show request method, URL, and response time.
-  add client env to .gitingnore.
-  downgrade jest to version 26.6.0 in package.json and package-lock.json to match the version in create-react-app.

Ahmed-aed committed:
create sign in route
related #19

Ahmed-aed committed:
create sign in test
related #19

Ahmed-aed committed:
fix unhanded errors in server

mFlaifel committed:
update signup and signinroute, add auth middle ware
relates #19 #23

mFlaifel committed:
fix update profile test after updating fake data

Ahmed-aed committed:
fixes requier and expots of signUpHandler and signInHandler
fixes #19 #23

Ahmed-aed committed:
resolving conflict

Sondos-Alalami commented on server/controllers/common/signInHandler.js:
```diff
@@ -0,0 +1,42 @@
+    })
+    .then(() => signIn(email))
+    .then(({ rowCount, rows }) => {
+      if (rowCount === 0) {
+        throw throwError(409, 'this email is not registered');
+      } else {
```
The status here should be 400, bad request

Sondos-Alalami commented on server/controllers/common/signInHandler.js:
```diff
@@ -0,0 +1,42 @@
+      delete clientData.role;
+      clientData.password = 'hidden for your security ';
+      res
+        .cookie('token', token, { httpOnly: true })
+        .json({ status: 201, message: 'sign in successful', clientData });
+    })
```
The status here may be 200, because no new resource has been created.

Sondos-Alalami commented on server/utilities/Validation /signUp.js:
```diff
@@ -1,16 +1,25 @@
-    first_name: Joi.string().required(),
-    last_name: Joi.string().required(),
-    password: Joi.string().alphanum().min(6).required(),
+    firstName: Joi.string().required(),
+    lastName: Joi.string().required(),
```
You can add min() and max() here for both first and last name, the name should not be too small or too long.

Sondos-Alalami commented on server/database/queries/signIn.js:
```diff
@@ -0,0 +1,11 @@
+const dbConnection = require('../connection');
+
+const signIn = (email) => {
+  const sql = {
+    text: 'SELECT * FROM users where email = $1 ;',
+    values: [email],
+  };
+  return dbConnection.query(sql);
+};
+
+module.exports = signIn;
```
This query is not necessary , you can use getClientDataByEmail and add the password, or if you want to keep it this way, it is better to name the function checkEmail for example, instead of signIn

Sondos-Alalami commented on server/controllers/common/signInHandler.js:
```diff
@@ -0,0 +1,42 @@
+      }
+    })
+    .then(({ password: hashedPassword }) => bcrypt.compare(password, hashedPassword))
+    .then((success) => {
+      if (!success) {
+        throw throwError(409, 'please check your email or password');
+      }
```
The status here should be 400, bad request

mFlaifel committed:
update status code
modify validation for string to include min and max
update the names of queries
relates #19

mFlaifel committed:
solve conflict with main

mFlaifel committed:
add role to clientData
relates #19

mFlaifel committed:
add role to clientData
relates #19

mFlaifel committed:
update branch to include remote branch updates
relates #19
